contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,248 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
 
-A new CRI API, `GetRuntimeConfigInfo` will be added. Kubelet will use this API:
-- To verify if user-namespace remapping is enabled at runtime. If found disabled, kubelet will fail to start
-- To determine the default user-namespace range at the runtime, starting UID of which is mapped to the UID '0' of the container.
+`HostUserNamespace` field will be gated behind `HostUserNamespace` feature-gate. To set it to `true` or `false`, feature-gate will have to be enabled. If feature-gate not enabled or feature-gate enabled but any value is not set, `nil` will be the default value. 
 
-### Volume Permissions
-Kubelet will change the file permissions, i.e chown, at `/var/lib/kubelet/pods` prior to any container start to get file permissions updated according to remapped UID and GID.
-This proposal will work only for local volumes and not with remote volumes such as NFS.
+### Effect on existing Security Context fields
 
-### How to disable `NodeUserNamespace` for a specific pod
-This can be done in two ways:
-- **Alpha:** Implicitly using host namespace for the pod containers
-This support is already present (currently it seems broken, will be fixed) in Kubernetes as an experimental functionality, which can be enabled using `feature-gates=”ExperimentalHostUserNamespaceDefaulting=true”`.
-If Pod-Security-Policy is configured to allow the following to be requested by a  pod, host user-namespace will be enabled for the container:
-  - host namespaces (pid, ipc, net)
-  - non-namespaced capabilities (mknod, sys_time, sys_module)
-  - the pod contains a privileged container or using host path volumes.
-  - https://github.com/kubernetes/kubernetes/commit/d0d78f478ce0fb9d5e121db3b7c6993b482af82c#diff-a53fa76e941e0bdaee26dcbc435ad2ffR437 introduced via https://github.com/kubernetes/kubernetes/commit/d0d78f478ce0fb9d5e121db3b7c6993b482af82c.
+* **RunAsUser**: UID to run the entrypoint of the container process. After usernamespace support in kubelet, if Kubelet discovers that usernamespace remapping is enabled at runtime, this will be the UID in the container usernamespace. In such a case, if no mapping(corresponding host UID) is found in the mapping range, pod will be failed admission at Kubelet. If kubelet discovers that usernamespace remapping is not supported at runtime, no validation check will be performed. 
+* **RunAsGroup**: same as **RunAsUser** but for gid mapping ranges.
+* **RunAsNonRoot**: Today, kubelet fails to start the container if image is found to be having UID 0. After usernamespace support in kubelet, kubelet should ignore this check if runtime is found to be running with usernamespace remapping enabled. Kubelet should verify image UID only when it is learned that usernamespace remapping is either not supported or not enabled at runtime.
```

Comment:
> kubelet should ignore this check if runtime is found to be running with usernamespace remapping enabled

I'm not sure I agree with this. There might still be use cases for requiring a non-root users in a user namespace (root users do have elevated privileges within the namespace, e.g. CAP_DAC_OVERRIDE). It also makes auditing difficult, especially in heterogeneous clusters.

Reply:
Agree with your point. `RunAsNonRoot`, just like `RunAsUser` and `RunAsGroup`, should be conveying meaning inside container usernamespace and thus existing check should be performed. Will update.